File: backend/internal/bulkcreate/fill.go
// Pure data-merge core (port of @hc/templates fill.ts), operating on the opaque
// DesignFile JSON (map[string]any). Text fields set their node's first run text;
// image fields point the node's image source at a URL. The source design is
// never mutated (applyFill clones first).
package bulkcreate

import (
	"encoding/json"
	"fmt"
)

// FillValue is one field's value: text fields use Text, image fields use ImageURL.
type FillValue struct {
	Text     string `json:"text,omitempty"`
	ImageURL string `json:"imageUrl,omitempty"`
}

// FillValues is a fill row keyed by fillable-field nodeId.
type FillValues map[string]FillValue

// Field is a fillable-field descriptor (matches the SDK FillableFieldSummary).
type Field struct {
	NodeID      string         `json:"nodeId"`
	Kind        string         `json:"kind"`
	Label       string         `json:"label"`
	Hint        string         `json:"hint,omitempty"`
	Constraints map[string]any `json:"constraints,omitempty"`
}

func cloneFile(file map[string]any) map[string]any {
	b, _ := json.Marshal(file)
	var out map[string]any
	_ = json.Unmarshal(b, &out)
	return out
}

func asArr(v any) []any { a, _ := v.([]any); return a }
func asObj(v any) map[string]any {
	m, _ := v.(map[string]any)
	return m
}

// findNode walks pages -> children (recursing into children, mask child, boolean
// operands) and returns the node with id, or nil.
func findNode(file map[string]any, id string) map[string]any {
	var found map[string]any
	var visit func(n map[string]any)
	visit = func(n map[string]any) {
		if found != nil || n == nil {
			return
		}
		if s, _ := n["id"].(string); s == id {
			found = n
			return
		}
		for _, c := range asArr(n["children"]) {
			visit(asObj(c))
		}
		if t, _ := n["type"].(string); t == "mask" {
			if ch := asObj(n["child"]); ch != nil {
				visit(ch)
			}
		}
		if t, _ := n["type"].(string); t == "boolean" {
			for _, op := range asArr(n["operands"]) {
				visit(asObj(op))
			}
		}
	}
	for _, p := range asArr(file["pages"]) {
		for _, root := range asArr(asObj(p)["children"]) {
			visit(asObj(root))
		}
	}
	return found
}

// fillTextNode sets a text node's copy, handling both the Paragraph (content[0]
// carries runs) and TableCell (content is the run array) shapes.
func fillTextNode(node map[string]any, text string) bool {
	content := asArr(node["content"])
	if len(content) == 0 {
		return false
	}
	first := asObj(content[0])
	if first == nil {
		return false
	}
	if runs := asArr(first["runs"]); runs != nil {
		if len(runs) == 0 {
			return false
		}
		run := asObj(runs[0])
		newRun := map[string]any{}
		for k, v := range run {
			newRun[k] = v
		}
		newRun["text"] = text
		first["runs"] = []any{newRun}
		node["content"] = []any{first}
		return true
	}
	if _, ok := first["text"]; ok {
		newFirst := map[string]any{}
		for k, v := range first {
			newFirst[k] = v
		}
		newFirst["text"] = text
		node["content"] = []any{newFirst}
		return true
	}
	return false
}

// fillImageNode points an image node (or a node's first image fill) at imageURL.
func fillImageNode(node map[string]any, imageURL string) bool {
	src := map[string]any{"assetId": imageURL, "naturalWidth": 0, "naturalHeight": 0}
	if t, _ := node["type"].(string); t == "image" {
		node["source"] = src
		delete(node, "crop")
		return true
	}
	for _, f := range asArr(node["fills"]) {
		fm := asObj(f)
		if fm == nil {
			continue
		}
		if ft, _ := fm["type"].(string); ft == "image" {
			fm["source"] = src
			delete(fm, "crop")
			return true
		}
	}
	return false
}

// ApplyFill substitutes a row of values into a clone of file. ids are preserved
// (a single autofill stays the same design); the input is never mutated.
func ApplyFill(file map[string]any, values FillValues) map[string]any {
	clone := cloneFile(file)
	for nodeID, value := range values {
		hasText := value.Text != ""
		hasImage := value.ImageURL != ""
		if !hasText && !hasImage {
			continue
		}
		node := findNode(clone, nodeID)
		if node == nil {
			continue
		}
		if hasText {
			fillTextNode(node, value.Text)
		}
		if hasImage {
			fillImageNode(node, value.ImageURL)
		}
	}
	return clone
}

// ValidateFillRow checks a row against the fields' constraints, returning the
// first problem (required presence + text maxChars; image aspect is deferred).
func ValidateFillRow(fields []Field, values FillValues) (bool, string) {
	for _, field := range fields {
		v, has := values[field.NodeID]
		present := has && (v.Text != "" || v.ImageURL != "")
		req, _ := field.Constraints["required"].(bool)
		if req && !present {
			return false, fmt.Sprintf("%s is required", field.Label)
		}
		if field.Kind == "text" {
			if mc, ok := numConstraint(field.Constraints["maxChars"]); ok && v.Text != "" && len([]rune(v.Text)) > mc {
				return false, fmt.Sprintf("%s exceeds %d characters", field.Label, mc)
			}
		}
	}
	return true, ""
}

func numConstraint(v any) (int, bool) {
	switch n := v.(type) {
	case float64:
		return int(n), true
	case int:
		return n, true
	}
	return 0, false
}

File: packages/templates/src/fill.ts
// Data merge / fill. Given a design (or
// template) that declares fillable fields, substitute a row of values into a
// COPY of the design: text fields set the text of their node's runs, image
// fields set the node's image source. The source design is never mutated.
//
// Bulk create (one design per dataset row) calls applyFill per row over a fresh
// deep copy; single-design autofill applies one row into a clone. The live-data
// binding, formula engine, and chart/table data sources are deferred (slice a /
// later); this module is the pure variable-data substitution core.

import { childrenOf, type DesignFile, type ImageSource, type Node } from "@hc/schema";
import type { FillableField } from "./types";

type AnyRec = Record<string, unknown>;

/** One field's value in a fill row. A text field uses `text`; an image field
 *  uses `imageUrl` (resolved to an ImageSource on the node). Absent/empty means
 *  "leave the node's existing default". */
export interface FillValue {
  text?: string;
  imageUrl?: string;
}

/** A fill row, keyed by fillable-field `nodeId`. */
export type FillValues = Record<string, FillValue>;

function findNode(file: DesignFile, id: string): Node | null {
  let found: Node | null = null;
  const visit = (n: Node) => {
    if (found) return;
    if (n.id === id) {
      found = n;
      return;
    }
    for (const c of childrenOf(n)) visit(c);
    const rec = n as unknown as AnyRec;
    if (n.type === "mask" && rec.child) visit(rec.child as Node);
    if (n.type === "boolean" && Array.isArray(rec.operands)) for (const op of rec.operands as Node[]) visit(op as Node);
  };
  for (const page of file.pages) for (const root of page.children) visit(root);
  return found;
}

/**
 * Set a text node's copy to `text` (FR-7). Handles two `content` shapes:
 *  - A TextNode's content is `Paragraph[]` (each paragraph has `runs`). The
 *    first run of the first paragraph carries the new text (preserving its
 *    style); further runs/paragraphs are collapsed to a single styled run.
 *  - A TableCell's content is `TextRun[]` directly (no `runs`). The first run
 *    carries the new text (preserving its style); further runs are dropped.
 * Non-text nodes are ignored. Mutates the passed (already-cloned) node.
 */
function fillTextNode(node: Node, text: string): boolean {
  const rec = node as unknown as AnyRec;
  const content = rec.content as AnyRec[] | undefined;
  if (!Array.isArray(content) || content.length === 0) return false;
  const first = content[0];
  // Paragraph shape: content[0] carries a `runs` array.
  if (Array.isArray(first?.runs)) {
    const runs = first.runs as AnyRec[];
    if (runs.length === 0) return false;
    const keepStyle = runs[0].style;
    first.runs = [{ ...runs[0], text, style: keepStyle }];
    // Drop trailing paragraphs so the field reads as a single value.
    rec.content = [first];
    return true;
  }
  // TextRun[] shape (table cell): content itself is the run array.
  if (first && typeof first === "object" && "text" in first) {
    rec.content = [{ ...first, text }];
    return true;
  }
  return false;
}

/**
 * Point an image node (or image-fill carrying node) at `imageUrl` (FR-11). The
 * URL is recorded as the source `assetId` (the engine resolves an assetId to a
 * blob; a URL is a degenerate-but-valid reference and self-hosters can resolve
 * it directly). Natural dimensions are unknown for a bare URL, so they reset to
 * 0 and the renderer falls back to the node's box. Mutates the cloned node.
 */
function fillImageNode(node: Node, imageUrl: string): boolean {
  const rec = node as unknown as AnyRec;
  const nextSource: ImageSource = { assetId: imageUrl, naturalWidth: 0, naturalHeight: 0 };
  if (node.type === "image") {
    rec.source = nextSource;
    delete rec.crop; // a new source invalidates the old normalized crop
    return true;
  }
  // A frame/shape with an image fill: replace the first image fill's source.
  const fills = rec.fills as AnyRec[] | undefined;
  if (Array.isArray(fills)) {
    for (const f of fills) {
      if (f?.type === "image") {
        f.source = nextSource;
        delete f.crop;
        return true;
      }
    }
  }
  return false;
}

export interface ApplyFillResult {
  /** The filled design (a deep clone; the input is never mutated). */
  file: DesignFile;
  /** Field node ids that received a value. */
  filled: string[];
  /** Field node ids whose value was provided but whose node was missing. */
  missing: string[];
}

/**
 * Substitute a row of values into a COPY of `file` (FR-8, FR-9). Each entry is
 * keyed by a fillable field's `nodeId`: a `text` value sets the node's text, an
 * `imageUrl` sets its image source. Fields with no value (or an empty string)
 * keep the design's existing default. The source design is never mutated; ids
 * are preserved (a single autofill stays the same design). Pure.
 *
 * Bulk create deep-copies (fresh ids) first, then applies a row's values to the
 * copy; pass the already-copied file here so this stays a focused substitution.
 */
export function applyFill(file: DesignFile, values: FillValues): ApplyFillResult {
  const clone: DesignFile = JSON.parse(JSON.stringify(file));
  const filled: string[] = [];
  const missing: string[] = [];

  for (const [nodeId, value] of Object.entries(values)) {
    if (!value) continue;
    const hasText = typeof value.text === "string" && value.text.length > 0;
    const hasImage = typeof value.imageUrl === "string" && value.imageUrl.length > 0;
    if (!hasText && !hasImage) continue; // no value: leave default

    const node = findNode(clone, nodeId);
    if (!node) {
      missing.push(nodeId);
      continue;
    }
    let applied = false;
    if (hasText) applied = fillTextNode(node, value.text!) || applied;
    if (hasImage) applied = fillImageNode(node, value.imageUrl!) || applied;
    if (applied) filled.push(nodeId);
    else missing.push(nodeId);
  }

  return { file: clone, filled, missing };
}

/**
 * Validate a whole fill row against the fields' constraints (FR-8 mapping
 * validation). Returns the first problem found, or ok. Image aspect/required
 * presence beyond a non-empty URL is not knowable here (dimensions are fetched
 * at ingest), so only text constraints and required-presence are enforced.
 */
export function validateFillRow(
  fields: FillableField[],
  values: FillValues,
): { ok: boolean; reason?: string } {
  for (const field of fields) {
    const v = values[field.nodeId];
    const c = field.constraints;
    const present =
      !!v && ((typeof v.text === "string" && v.text.length > 0) || (typeof v.imageUrl === "string" && v.imageUrl.length > 0));
    if (c?.required && !present) return { ok: false, reason: `${field.label} is required` };
    if (field.kind === "text" && c?.maxChars !== undefined && v?.text && v.text.length > c.maxChars) {
      return { ok: false, reason: `${field.label} exceeds ${c.maxChars} characters` };
    }
  }
  return { ok: true };
}
